[ENT-486] Programs Landing Page

UmanShahzad opened the pull request:
**Description:** This PR adds a Programs landing page, akin to the course enrollment landing page.

**JIRA:** [ENT-486](https://openedx.atlassian.net/browse/ENT-486)

**Dependencies:** None any longer.

**Merge deadline:** Asap.

**Testing instructions:**

You will need to utilize the LMS, discovery, and ecommerce for properly testing this, as well as fake data to actually see anything. Or this can be moved to the sandbox and be tested there.

#### Setup

1. Make authoring organizations on course-discovery.
1. Make courses (audit and verified track) and course runs, and use any authoring organization if applicable. Make them appear on course-discovery.
1. Make a program containing anything you created above, if applicable (and note the UUID). Make it one-click purchasable.
1. Make a program API config (actually, it's just a "program config", but we aren't changing the name yet on the model) at LMS path `/admin/programs/programsapiconfig/`. Make sure it's enabled.
1. Make SKU data associated with the course(s) you made above on ecommerce.
1. Make an enterprise customer. Note the UUID.

#### Initial test of GET, POST, and UI

1. Login to the LMS as any user who is not enrolled in the course you created.
1. Navigate to `/enterprise/{enterprise_uuid}/program/{program_uuid}/enroll/`. Replace `{enterprise_uuid}` and `{program_uuid}` with the enterprise's UUID and the program's UUID that you noted above, respectively.
1. See something roughly like this page (information present depends on your configuration of fake data):

![selection_002](https://user-images.githubusercontent.com/10018065/30072566-2a08a348-9284-11e7-8d72-f9c2f94f4b50.png)

1. Click the Confirm Program button.
1. Ensure that you see the DSC page.
1. Decline consent by clicking "No, take me back".
1. Ensure that you're redirected back to the program enrollment landing page, and see the message at the top, as such:

![selection_001](https://user-images.githubusercontent.com/10018065/30072499-f2f39048-9283-11e7-9f2e-033de5048cbf.png)

1. Click the Confirm Program button again.
1. This time, grant data sharing consent.
1. Ensure that you see the basket page with the list of courses (or just 1) that you have in the program.
1. Ensure that the price matches what was advertised on the program enrollment landing page that you just navigated away from.
1. Ensure that if a discount was present, that discount is applied in the basket.
1. Fill in basket information to get yourself enrolled in the program.
1. Ensure you are redirected to the Program's detail page (`/dashboard/programs/{program_uuid}/`) upon enrollment.

#### Making sure GET redirects us to the Program's detail page

1. After initial enrollment is successful, navigate to `/enterprise/{enterprise_uuid}/program/{program_uuid}/enroll/` once more, replacing UUID placeholders as necessary.
1. Notice you are redirected to the Program's detail page (`/dashboard/programs/{program_uuid}/`) instead of seeing the program enrollment landing page.

#### Making sure you see adjusted text & price when enrolled in at least 1 course, but not all

1. If you only had 1 course in the program, create more and put them in the program. If you already had multiple, manually unenroll yourself from one or more of the courses, but not all.
1. Navigate to `/enterprise/{enterprise_uuid}/program/{program_uuid}/enroll/`.
1. See that you are now on the program enrollment landing page.
1. See that for each course you were still enrolled in, it is striked out and next to it reads "**(enrolled)**".
1. See that for each course you *weren't* enrolled in, it appears as normal.
1. See that the price is adjusted to match the total of only those courses you're not already enrolled in (consider any discounts, too).

#### Making sure you see adjusted text & price when enrolled in a course in the program but without having paid for the certificate

1. Unenroll in at least 1 course in the program manually.
1. Re-enroll in the course, but this time on the audit track. You can use the enterprise course enrollment landing page if you like.
1. Navigate to `/enterprise/{enterprise_uuid}/program/{program_uuid}/enroll/`.
1. See that you are now on the program enrollment landing page.
1. See that for each course you were still enrolled in in audit mode, below the course reads "**(already enrolled, must pay for certificate)**".
1. See that for each course you *weren't* enrolled in, it appears as normal.

**Merge checklist:**

- [ ] Check that the versions of the requirements in the `platform-master.in` file match edx-platform.
- [x] New requirements are in the right place (`base.in` if only used in enterprise; in the correct `platform-****.in` files if they're hosted in edx-platform)
- [x] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [ ] All reviewers approved
- [x] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [x] Documentation updated (not only docstrings)
- [ ] Commits are (reasonably) squashed
- [x] Translations are updated
- [ ] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)
- [ ] edx-platform PR (be sure to include edx-platform requirements upgrades that were present in this PR)

UmanShahzad commented on enterprise/api_client/enterprise.py:
```diff
@@ -0,0 +1,93 @@
+# -*- coding: utf-8 -*-
```
This was originally from #169, but I needed it for this so copied it here and modified it to be generic + have the needed additions for programs.

UmanShahzad commented on enterprise/static/enterprise/enterprise_landing_page.css:
```diff
@@ -0,0 +1,503 @@
+* {
```
It's copy and pasted from the other CSS files because they will be sharing style a lot -- ignore for now.

UmanShahzad commented on enterprise/templates/enterprise/enterprise_program_enrollment_page.html:
```diff
@@ -0,0 +1,172 @@
+{% extends 'enterprise/base.html' %}
```
It's copy and pasted from the other template files because they will be sharing format a lot -- ignore for now.

UmanShahzad commented on enterprise/utils.py:
```diff
@@ -533,3 +536,44 @@ def get_enterprise_customer_or_404(enterprise_uuid):
+        response = endpoint.get(**querystring)
+        results += response.get('results', [])
+        next_page = response.get('next')
+    return results
```
Copied from #169 because I needed them for the Enterprise API.

UmanShahzad commented on test_settings.py:
```diff
@@ -129,6 +129,7 @@ def root(*args):
 ENTERPRISE_CUSTOMER_LOGO_IMAGE_SIZE = 512   # Enterprise logo image size limit in KB's
 
 ENTERPRISE_COURSE_ENROLLMENT_AUDIT_MODES = ['audit', 'honor']
+ENTERPRISE_COURSE_ENROLLMENT_VERIFIED_MODES = ['verified', 'professional']
```
Will have to make a PR on `edx/configuration` for this and add it to `edx/edx-platform` as well, I believe. Not even sure if this is correct, though!

UmanShahzad commented:
@itsjeyd I tried basing on `uman/fix-new-endpoints`, which is the branch that pulled out a subset of the changes in this PR to put into a separate PR to get it over the line faster, but the changeset doesn't seem to reflect the work overlap, so I'm a bit sorry about that! This PR also requires those changes so it's a little awkward.

edx-webhook commented:
Thanks for the pull request, @UmanShahzad! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=171&repo=edx%2Fedx-enterprise"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

UmanShahzad commented:
@haikuginger @itsjeyd There's test coverage now, minus coverage for some code I borrowed from #169 because I needed it here -- when test coverage is added there I'll pick up those commits and put 'em here.

UmanShahzad commented on enterprise/api/v1/mixins.py:
```diff
@@ -0,0 +1,100 @@
+"""
```
:+1:

UmanShahzad commented on enterprise/api/v1/mixins.py:
```diff
@@ -0,0 +1,100 @@
+        courses = []
+        for course in self.data[course_container_key]:
+            courses.append(
+                self.update_course(course, enterprise_customer, enterprise_context)
```
Indeed, I had to reuse this functionality so I decided to drop it in a mixin because it was just a bunch of functions.

UmanShahzad commented:
@haikuginger I'm in the same boat as you and really wish the dependencies got sorted out first before we pushed forward with review for this ticket, TBH, because otherwise it's too damn difficult.

UmanShahzad commented:
@haikuginger Just for your note, rebased from 1e26c7e21e0eafd5dc30d02341726b39d0c25512 on `master`. Also I know you may not have gotten around to explicitly testing the new endpoint introduced in this, but since we'll inevitably be testing it for the manual testing of the overall PR (which is now into its 2nd week of new work), are the current changes good enough for a +1 from you?

UmanShahzad commented:
I gave the endpoint changes from #170 a test and the endpoints work as we expect, all 4 of

- `enterprise-catalogs-list`
- `enterprise-catalogs-detail`
- `programs-list`
- `programs-detail`

I'm also currently working on the front-end ATM, and can confirm that the current change set with GET for our new view work as expected, i.e. prices are calculated, all the proper information is given and shown. Now I'm just working on the CSS, and the POST function, and we should be good for a thorough review.

UmanShahzad commented:
In order to make solving the rebase conflicts easier for myself, I'm squashing these commits down to something more reasonable.

UmanShahzad commented:
Squashed from a commit that's not actually here... sorry. The last commit that was showing here was b130f07dcaece44b2ede94b30ce06b1a18136a61.

When #170 gets merged, I'll rewrite the commit history to tell a better tale.

itsjeyd commented:
@UmanShahzad 

> ... are the current code changesets good enough for a +1 from you?

Just a quick note that you don't need to wait for an explicit +1 from @haikuginger here. We should stick to using +1 for final approval. Tracking +1s on a large PR can become confusing/cumbersome otherwise (been there), and make +1s lose their meaning (which has been, "I'm happy with both code and behavior introduced in this PR, and consider it ready for merge").

UmanShahzad commented:
@itsjeyd Sure, I get what you mean -- it doesn't have to be an explicit +1 in that case, but just an acknowledgement that "yeah, this part looks good now".

UmanShahzad commented:
I've rebased on the DSC merge. I have some coverage missing but I'll fill it up whenever #173 is ready for merge, so that I can write all the remaining code in one-go.

UmanShahzad commented:
@haikuginger Rebased on your work, and added the necessary DSC checks for POST. Only thing left: add coverage and test this thing.

UmanShahzad commented:
@haikuginger @itsjeyd @douglashall @brittneyexline @zrana Added last bit of test coverage for this changeset (however the code from #171 shows as needing some coverage), added data sharing consent check in POST, and self-tested it a bit -- works as expected.

Please give this a test run if not a code review. I'd say we'd be ready for code review when #171 and #173 get merged, the only remaining dependencies.

UmanShahzad commented:
Squashed to 1 commit and rebased on master so the rebase is easier. I will wait for #170 to merge before I rewrite the commit history.

UmanShahzad commented:
The test failures are flaky from `master`. We need to really handle them soon.

UmanShahzad commented:
I've taken care of these flaky tests by adding the `flaky` decorator, which is also used in `edx-platform`.

UmanShahzad commented:
I've rebased this on `master`, but now going to re-write the history into smaller chunks and then push.

UmanShahzad commented:
@haikuginger @douglashall @brittneyexline Rebased on `master` and re-wrote the history. Ready for review! Not on the sandbox yet, though.

UmanShahzad commented:
@douglashall Woops, thanks, totally forgot about that TODO :+1:

UmanShahzad commented on enterprise/views.py:
```diff
@@ -952,3 +964,376 @@ def get(self, request, enterprise_uuid, course_id):
+            course_id = program_course_run['key']
+            # Track courses so we can calculate the price later.
+            course_modes = enrollment_client.get_course_modes(course_id)
+            course_mode_for_calculation = [mode for mode in course_modes if mode['slug'] == 'verified'][0]
```
Pretty sure I'm doing something wrong here, because I think I need to include 'professional', but not sure. Please advise!

UmanShahzad commented:
@douglashall The test failures due to the commit are due to a difference in having an ending `/`. I can fix that, but will do so a little later: working on setting up the sandbox environment for testing.

UmanShahzad commented on test_utils/__init__.py:
```diff
@@ -90,7 +90,10 @@ def update_program_with_enterprise_context(program):
-        '/enterprise/{}/program/{}/enroll'.format(FAKE_UUIDS[0], program['uuid'])
+        reverse(
+            'enterprise_program_enrollment_page',
+            kwargs = {'enterprise_uuid': FAKE_UUIDS[0], 'program_uuid': program['uuid']}
```
Darn, I left a space. I'll fix it on some future squash.

UmanShahzad replied:
Nevermind fixing it now.

UmanShahzad commented:
Fixed and squashed. This is now on the sandbox at https://business.sandbox.edx.org/enterprise/1b542b06-d353-45e6-9867-9579d9e12745/program/52ad909b-c57d-4ff1-bab3-999813a2479b/enroll/ -- however I only put 1 course in there at the moment. I'll put more and make discounts show, etc..

douglashall commented on enterprise/views.py:
```diff
@@ -952,3 +964,376 @@ def get(self, request, enterprise_uuid, course_id):
+            'program_price': program_price,
+            'program_discounted_price': program_discounted_price,
+            'courses': program_courses,
+        }
```
@UmanShahzad There is prior art for calculating the learner-specific program information we need to render the program enrollment page.

https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/programs/utils.py#L454

We may want to consider refactoring some of this existing code to make it usable by our enterprise code. We may also want to consider breaking our standard approach of using REST APIs for communicating with LMS code for performance sake.

UmanShahzad replied:
@douglashall Just to confirm -- are you asking to import these utilities from the platform and re-use them here in place of the current changeset, or to refactor this changeset to be more like the utilities in the platform, and to be reusable throughout the edx-enterprise repo?

As for not using REST APIs, I'm wondering what else you have in mind to replace that -- in many cases, we simply need a lot of the information that the endpoint provides.

douglashall replied:
My initial thinking was to refactor the edx-platform code linked above to allow us to import a single function that would do the "one-click purchase eligibility" calculation and return it to us, but now I'm thinking we just import `ProgramDataExtender`, pass it the program data dictionary, and let it do it's thing, instead of recreating the logic here.

As for performance, I think it would be better to avoid several API roundtrips to render a single page. We could create a new function on the ecommerce API client that lives in edx-enterprise which wraps the call to `ProgramDataExtender` and returns the "extended" program data.

UmanShahzad replied:
Ah, I understand what you mean. So we can do that, but 2 things:

1. This would entail a lot of code changes, especially in the tests (and the test code is large), and considering it's a refactor and we're (I believe) under pressure with a deadline, we should maybe consider focusing most on testing and let the refactor happen in the follow-up story (i.e. adding details) or in a separate story from these both, whatever seems appropriate.
1. This makes me wonder why this repository isn't just a part of the platform directly if solutions are best done by just  :laughing:

douglashall commented on enterprise/static/enterprise/enterprise_program_enrollment_page.css:
```diff
@@ -0,0 +1,268 @@
+  background-color: #f2dede;
+  border-color: #ebccd1;
+  color: #a94442;
+}
```
How much of this is copy/paste from enterprise_course_enrollment_page.css? Is it worth consolidating to a base.css which contains the common styles and can be included on both pages?

UmanShahzad replied:
The `#messages` and `.btn-confirm`, and about the first ~86 lines are copy pasted, with some changes within those lines. The `enterprise_course_enrollment_page.css` contains additional styles that were of course deleted here. I was thinking to combine them, but I was also aware that in the future there would be stories about consolidating the templates with fragments (or something), so I didn't bother too invest effort into doing any consolidation work here.

douglashall commented on enterprise/messages.py:
```diff
@@ -13,34 +14,108 @@
-                support_link=configuration_helpers.get_value('ENTERPRISE_SUPPORT_URL')
+                support_link=configuration_helpers.get_value(
+                    'ENTERPRISE_SUPPORT_URL',
+                    getattr(settings, 'ENTERPRISE_SUPPORT_URL')  # Remove the `getattr` when setting is upstreamed.
```
This all looks right, except you could provide a default value of `''` or `None` to `getattr`until we have things upstreamed so we don't get the 500 error.

UmanShahzad replied:
I'll do that before I log so we stop getting errors and update the sandbox.

haikuginger commented on enterprise/views.py:
```diff
@@ -952,3 +964,376 @@ def get(self, request, enterprise_uuid, course_id):
+                the Enterprise..
+        """
+        verify_edx_resources()
+        enterprise_customer, program_details, program_enrollment_details = self.get_base_details(
```
I believe we need to do this _after_ we do the EnterpriseCustomerUser creation; otherwise we won't get accurate entitlement details for the user based on Enterprise linkage from the Ecommerce API. That's the whole reason behind making the view nonatomic in the first place.

UmanShahzad replied:
:+1:

haikuginger commented on enterprise/views.py:
```diff
@@ -952,3 +964,376 @@ def get(self, request, enterprise_uuid, course_id):
+        )
+
+        # Create a link between the user and the enterprise customer if it does not already exist.
+        with transaction.atomic():
```
As above in the GET handler.

UmanShahzad replied:
:+1:

haikuginger commented on tests/test_admin/test_forms.py:
```diff
@@ -138,8 +138,8 @@ def test_clean_empty_email_and_no_file(self, email):
-        UserFactory(username="user1", email="user1@example.com")
-        UserFactory(username="user2", email="user2@example.com")
+        UserFactory(username="user1", email="user1@example.com", id=1)
+        UserFactory(username="user2", email="user2@example.com", id=2)
```
Why do we need to pass the IDs to the factories?

haikuginger replied:
Got it! I'm in favor of the change, then.

haikuginger commented on tests/test_enterprise/api_client/test_enterprise.py:
```diff
@@ -51,6 +59,54 @@ def _assert_num_requests(self, expected_count):
+        cache.set(cache_key, {'fake': 'response'}, settings.ENTERPRISE_API_CACHE_TIMEOUT)
+        client = enterprise_api.EnterpriseApiClient(self.user)
+        response = client._load_data('resource')  # pylint: disable=protected-access
+        cached_api_response = cache.get(cache_key)
```
Instead of getting the comparison value out of the cache, let's save the value we cache above into a variable, and then compare against the returned value.

UmanShahzad replied:
:+1:

UmanShahzad commented on enterprise/views.py:
```diff
@@ -952,3 +964,376 @@ def get(self, request, enterprise_uuid, course_id):
+        'program_not_eligible_for_one_click_purchase_text': _('Program not eligible for one-click purchase.'),
+    }
+
+    def get_program_prices(self, program_courses, user):
```
We are considering refactoring this, yes, to reuse logic from the platform, according to this comment thread: https://github.com/edx/edx-enterprise/pull/171#discussion_r137093788 -- that's next on the agenda. Any thoughts on that though before I start?

UmanShahzad commented on enterprise/api_client/enterprise.py:
```diff
@@ -19,37 +24,70 @@ class EnterpriseApiClient(JwtLmsApiClient):
+        )
+
+    @JwtLmsApiClient.refresh_token
+    def _load_data(
```
So this code was changed originally to accommodate using the (now gone) `/enterprise/api/v1/programs/{uuid}` endppoint we had from ENT-341, to get program details. I had changed this code in this PR to be more generic to allow me to easily query that. But now that it's no longer necessary these abstractions seem like they were done for the fun of it. I agree with your point of view on it, though, and will make some fixes.

UmanShahzad commented on enterprise/api_client/enterprise.py:
```diff
@@ -19,37 +24,70 @@ class EnterpriseApiClient(JwtLmsApiClient):
-            resource=api_resource_name,
-            enterprise_uuid=enterprise_customer.uuid,
-            traverse=traverse
+        return self._load_data(
```
:+1:

UmanShahzad commented on enterprise/utils.py:
```diff
@@ -555,3 +557,37 @@ def traverse_pagination(response, endpoint):
+
+    :param price: The price amount.
+    :param currency: The currency for the price.
+    :return: A formatted price string or decimal value if no currency, i.e. '$10', '$10.52', or 10 & 10.52.
```
The use case was really only for the program price calculation, but since we're going to refactor that, this may just disappear entirely.

UmanShahzad commented on enterprise/utils.py:
```diff
@@ -555,3 +557,37 @@ def traverse_pagination(response, endpoint):
     return results
+
+
+def ungettext_min_max(singular, plural, min_val, max_val, info_text):
```
:+1:

UmanShahzad commented on enterprise/api_client/ecommerce.py:
```diff
@@ -41,19 +41,51 @@ def __init__(self, user):
+            )
+        }
+
+    def calculate_basket_price(self, skus):
```
Yep, I had done this at some earlier point in time and refactored and no longer needed it. I will delete the unused stuff.

UmanShahzad commented:
@haikuginger Thanks for the review -- as I work on the refactor according to https://github.com/edx/edx-enterprise/pull/171#discussion_r137093788, I'll make these changes, because there is some overlap.

UmanShahzad commented:
@haikuginger I've addressed your review, except for your note about `format_prices` and the possible duplication of logic around calculating price info -- I will be addressing those in the refactor to use the program utilities from the platform.

UmanShahzad commented:
f3543a07f28674796f7ad7669b336af0e11a5d01

UmanShahzad commented:
@douglashall So I gave the program extender utility from the platform a test run on my local program, and here's the data I got:

```json
{
	"applicable_seat_types": [
		"verified",
		"professional",
		"credit"
	],
	"authoring_organizations": [
		{
			"certificate_logo_image_url": null,
			"description": "",
			"homepage_url": null,
			"key": "edX",
			"logo_image_url": null,
			"marketing_url": null,
			"name": "edX",
			"tags": [],
			"uuid": "2bb51703-3a1a-4c35-b9e3-315e53eb2f39"
		},
		{
			"certificate_logo_image_url": null,
			"description": "We are da org.",
			"homepage_url": "http://da.org",
			"key": "da-org",
			"logo_image_url": "http://da.org/boring/org/image.png",
			"marketing_url": null,
			"name": "Da Org",
			"tags": [],
			"uuid": "3cf09aef-e2e5-47d8-ae72-15124503ae4f"
		}
	],
	"banner_image": {
		"large": {
			"height": 480,
			"url": "http://edx.devstack.discovery:18381/media/media/programs/banner_images/0fd321b8-3f15-4541-8224-3cf081923e34-5dc618b04934.large.png",
			"width": 1440
		},
		"medium": {
			"height": 242,
			"url": "http://edx.devstack.discovery:18381/media/media/programs/banner_images/0fd321b8-3f15-4541-8224-3cf081923e34-5dc618b04934.medium.png",
			"width": 726
		},
		"small": {
			"height": 145,
			"url": "http://edx.devstack.discovery:18381/media/media/programs/banner_images/0fd321b8-3f15-4541-8224-3cf081923e34-5dc618b04934.small.png",
			"width": 435
		},
		"x-small": {
			"height": 116,
			"url": "http://edx.devstack.discovery:18381/media/media/programs/banner_images/0fd321b8-3f15-4541-8224-3cf081923e34-5dc618b04934.x-small.png",
			"width": 348
		}
	},
	"card_image_url": "http://globe-views.com/dcim/dreams/card/card-04.jpg",
	"corporate_endorsements": [],
	"courses": [
		{
			"course_runs": [
				{
					"advertised_start": null,
					"certificate_url": null,
					"course_url": "/courses/course-v1:edX+DemoX+Demo_Course/",
					"end": null,
					"enrollment_end": null,
					"enrollment_open_date": "Jan 01, 1900",
					"enrollment_start": null,
					"image": {
						"description": null,
						"height": null,
						"src": "http://edx.devstack.lms:18000/asset-v1:edX+DemoX+Demo_Course+type@asset+block@images_course_image.jpg",
						"width": null
					},
					"is_course_ended": false,
					"is_enrolled": false,
					"is_enrollment_open": true,
					"key": "course-v1:edX+DemoX+Demo_Course",
					"marketing_url": "course/edxdemoslug",
					"may_certify": false,
					"pacing_type": "instructor_paced",
					"seats": [
						{
							"credit_hours": null,
							"credit_provider": null,
							"currency": "USD",
							"price": "149.00",
							"sk": "8CF08E5",
							"type": "verified",
							"upgrade_deadline": "2018-08-20T22:07:16.102639Z"
						},
						{
							"credit_hours": null,
							"credit_provider": null,
							"currency": "USD",
							"price": "0.00",
							"sk": "68EFFFF",
							"type": "audit",
							"upgrade_deadline": null
						}
					],
					"short_description": null,
					"start": "2013-02-05T05:00:00Z",
					"status": "published",
					"title": "edX Demonstration Course",
					"type": "verified",
					"upgrade_url": null,
					"uuid": "9b560591-9b69-4a89-ae99-e9756ec4700a"
				}
			],
			"image": null,
			"key": "edX+DemoX",
			"owners": [
				{
					"key": "edX",
					"name": "edX",
					"uuid": "2bb51703-3a1a-4c35-b9e3-315e53eb2f39"
				}
			],
			"short_description": null,
			"title": "edX Demonstration Course",
			"uuid": "4e7820d8-03ef-42da-ba33-a4260ed568f8"
		},
		{
			"course_runs": [
				{
					"advertised_start": null,
					"certificate_url": null,
					"course_url": "/courses/course-v1:edX+E2E-101+course/",
					"end": "2018-12-31T00:00:00Z",
					"enrollment_end": "2018-12-31T00:00:00Z",
					"enrollment_open_date": "Jan 01, 2016",
					"enrollment_start": "2016-01-01T00:00:00Z",
					"image": {
						"description": null,
						"height": null,
						"src": "http://edx.devstack.lms:18000/static/01e4124377/images/pencils.jpg",
						"width": null
					},
					"is_course_ended": false,
					"is_enrolled": false,
					"is_enrollment_open": true,
					"key": "course-v1:edX+E2E-101+course",
					"marketing_url": null,
					"may_certify": false,
					"pacing_type": "instructor_paced",
					"seats": [
						{
							"credit_hours": null,
							"credit_provider": null,
							"currency": "USD",
							"price": "100.00",
							"sk": "DC4DC36",
							"type": "verified",
							"upgrade_deadline": "2019-04-19T00:00:00Z"
						},
						{
							"credit_hours": null,
							"credit_provider": null,
							"currency": "USD",
							"price": "0.00",
							"sk": "433CCC2",
							"type": "audit",
							"upgrade_deadline": null
						}
					],
					"short_description": null,
					"start": "2016-01-01T00:00:00Z",
					"status": "published",
					"title": "Manual Smoke Test Course 1 - Auto",
					"type": "verified",
					"upgrade_url": null,
					"uuid": "ba93b61e-fd37-45a1-a865-0de6920b1797"
				}
			],
			"image": null,
			"key": "edX+E2E-101",
			"owners": [
				{
					"key": "edX",
					"name": "edX",
					"uuid": "2bb51703-3a1a-4c35-b9e3-315e53eb2f39"
				}
			],
			"short_description": null,
			"title": "Manual Smoke Test Course 1 - Auto",
			"uuid": "dd2a786b-af5c-40bb-9e94-8c4db6ff0917"
		}
	],
	"credit_backing_organizations": [
		{
			"certificate_logo_image_url": null,
			"description": "",
			"homepage_url": null,
			"key": "edX",
			"logo_image_url": null,
			"marketing_url": null,
			"name": "edX",
			"tags": [],
			"uuid": "2bb51703-3a1a-4c35-b9e3-315e53eb2f39"
		}
	],
	"credit_redemption_overview": "Credit redemption overview.",
	"discount_data": {
		"currency": "USD",
		"discount_value": 10,
		"is_discounted": true,
		"total_incl_tax": 239,
		"total_incl_tax_excl_discounts": 249
	},
	"expected_learning_items": [
		"XBlocks",
		"SAPSF",
		"Software Development",
		"Learning items"
	],
	"faq": [],
	"full_program_price": 239,
	"hidden": false,
	"individual_endorsements": [],
	"is_learner_eligible_for_one_click_purchase": true,
	"is_program_eligible_for_one_click_purchase": true,
	"job_outlook_items": [],
	"languages": [
		"en-us"
	],
	"marketing_slug": "programmarketingslug1",
	"marketing_url": "professional-certificate/programmarketingslug1",
	"max_hours_effort_per_week": 8,
	"min_hours_effort_per_week": 4,
	"overview": "Gain the critical thinking skills and knowledge needed to become a more effective decision maker.",
	"price_ranges": [
		{
			"currency": "USD",
			"max": 149,
			"min": 100,
			"total": 249
		}
	],
	"skus": [
		"8CF08E5",
		"DC4DC36"
	],
	"staff": [],
	"status": "active",
	"subjects": [],
	"subtitle": "Program Subtitle",
	"title": "Evidence-Based Management",
	"transcript_languages": [
		"en-us"
	],
	"type": "Professional Certificate",
	"uuid": "0fd321b8-3f15-4541-8224-3cf081923e34",
	"variant": "full",
	"video": null,
	"weeks_to_complete": null,
	"weeks_to_complete_max": 5,
	"weeks_to_complete_min": 5
}
```

It looks pretty good! I'll reuse this and we should be able to drop a bunch of logic.

UmanShahzad commented:
@douglashall @haikuginger I refactored the code to use the `ProgramDataExtender` from the edx-platform. It made the code significantly easier to read imo. Functionality should not have changed at all based off of my local devstack testing.

UmanShahzad commented on enterprise/views.py:
```diff
@@ -1198,57 +1112,66 @@ def get_enterprise_program_enrollment_page(
+
+        one_click_purchase_eligibility = program_details['is_learner_eligible_for_one_click_purchase']
+        if not one_click_purchase_eligibility:
+            add_not_one_click_purchasable_message(request, enterprise_customer, program_title)
```
Sorry for moving code around, like here. Since I was refactoring, I tried to make it more organized, as well.

UmanShahzad commented on enterprise/views.py:
```diff
@@ -1198,57 +1112,66 @@ def get_enterprise_program_enrollment_page(
-                purchase_text=self.context_data['purchase_text'].format(
-                    purchase_action=purchase_action.lower(),
-                )
-            ),
```
I removed the "Free to try..." text as discussed in the meeting.

haikuginger commented on enterprise/views.py:
```diff
@@ -972,3 +990,271 @@ def get(self, request, enterprise_uuid, course_id):
+            '{count} Courses',
+            course_count,
+        ).format(count=course_count)
+        effort_info_text = ungettext_min_max(
```
Does not consider the case where one of these values might be missing; we should probably skip this item if we don't have data.

UmanShahzad replied:
I would assume we would have the data (these are pretty important parts of the page we're after), but yeah, I'll do this check.

UmanShahzad replied:
P.S. I ended up not skipping the item. I think we can do this in the follow-up ticket to save time here.

haikuginger commented on enterprise/views.py:
```diff
@@ -972,3 +990,271 @@ def get(self, request, enterprise_uuid, course_id):
+            program_details['min_hours_effort_per_week'],
+            program_details['max_hours_effort_per_week'],
+        )
+        length_info_text = ungettext_min_max(
```
As with the effort information text.

UmanShahzad replied:
:+1:

UmanShahzad commented on enterprise/views.py:
```diff
@@ -972,3 +990,271 @@ def get(self, request, enterprise_uuid, course_id):
+        try:
+            program_details = CourseCatalogApiServiceClient().get_program_by_uuid(program_uuid)
+        except (HttpClientError, ImproperlyConfigured) as error:
+            LOGGER.error('Failed to get program details for program with UUID %s due to: %s', program_uuid, str(error))
```
:+1: Yeah I have yet to rebase this on the #182 that I just now merged, and would have done something like that then.

UmanShahzad commented on enterprise/views.py:
```diff
@@ -972,3 +990,271 @@ def get(self, request, enterprise_uuid, course_id):
+        Render Enterprise-specific program enrollment page.
+        """
+        # Safely make the assumption that we can use the first authoring organization.
+        organization = program_details['authoring_organizations'][0]
```
As per the comment, I was told from @douglashall that we can assume the first is fine to use. We do something similar in the course enrollment view with `organization = course['owners'][0]`. Either way, I'll cover this case and return `{}` otherwise, and `.get()` things from it below..

UmanShahzad commented on test_utils/fake_enterprise_api.py:
```diff
@@ -30,25 +30,17 @@ def setUp(self):
+        return '{lms_root_url}{enterprise_api_uri}{params}'.format(
+            lms_root_url=settings.LMS_ROOT_URL,
+            enterprise_api_uri=reverse(resource, args=args),
+            params=('?' + urlencode(kwargs)) if len(kwargs) else '',  # pylint: disable=len-as-condition
```
Ah right! :+1:

UmanShahzad commented on enterprise/views.py:
```diff
@@ -972,3 +990,271 @@ def get(self, request, enterprise_uuid, course_id):
+            program_uuid=program_uuid,
+        )
+        if get_data_sharing_consent(
+                enterprise_customer_user.username,
```
Yep, I had the same impression but that's just how it is :P

UmanShahzad commented on enterprise/views.py:
```diff
@@ -972,3 +990,271 @@ def get(self, request, enterprise_uuid, course_id):
+        if program_data_sharing_consent.exists and not program_data_sharing_consent.granted:
+            add_consent_declined_message(request, enterprise_customer, program_title)
+
+        discount_data = program_details['discount_data']
```
Sure, will return `{}` with `.get`.

UmanShahzad commented on enterprise/views.py:
```diff
@@ -972,3 +990,271 @@ def get(self, request, enterprise_uuid, course_id):
+            # The user is already enrolled in the program, so redirect to the program's dashboard.
+            return redirect(LMS_PROGRAMS_DASHBOARD_URL.format(uuid=program_uuid))
+
+        basket_page = '{basket_url}?{skus}&bundle={program_uuid}'.format(
```
:+1:

UmanShahzad commented:
Rebased from 66c39d710fdf6d231340402e97b9e5499494ba52 on `master`. @haikuginger I've addressed the review as well.

UmanShahzad commented on enterprise/views.py:
```diff
@@ -1193,11 +1191,8 @@ def post(self, request, enterprise_uuid, program_uuid):
-                },
-                doseq=True
+                [tuple(['sku', sku]) for sku in program_details['skus']] +
+                [tuple(['bundle', program_uuid])]
```
This is to preserve order for tests ._.

haikuginger commented on enterprise/views.py:
```diff
@@ -933,3 +949,272 @@ def get(self, request, enterprise_uuid, course_id):
+            'enterprise_customer': enterprise_customer,
+            'organization_name': organization.get('name'),
+            'organization_logo': organization.get('logo_image_url'),
+            'organization_text': self.context_data['organization_text'].format(organization=organization['name']),
```
Crash with missing org isn't quite fixed; the `organization` doesn't have a `name` key if it's missing above. This is a relatively minor thing, but I want to make sure that we're building something that's not only suitable for edX.

UmanShahzad replied:
Ah, thanks :+1:

UmanShahzad commented:
@haikuginger Thanks! I've made the change, not sure how it slipped by :)

UmanShahzad commented:
@douglashall @brittneyexline Just need one of your approvals and we can go for it! For the 1 super tiny nit Christine mentioned (changing a static text slightly), it's in here already and probably not worth putting on the sandbox and going for another review cycle just for that.

UmanShahzad commented:
@douglashall Will squash and merge on your go!

douglashall commented:
No need to wait for me @UmanShahzad. Looks like you have what you need to merge! Nice work.
